Use UTF-8 Box Drawing for dist-repo-status

jmgate opened the pull request:
If `sys.stdout.encoding` supports UTF-8, use the box drawing characters to create a nicely formatted table for `dist-repo-status`.  If UTF-8 is not supported, fall back to the original ASCII table.

bartlettroscoe commented:
@jmgate, thanks for the PR.  I am curious to see what the improved formatting looks like.

I just tried this on my machine with Python 2.6.6 (the default on RHEL6) and it returned:

```
$ which python
/usr/bin/python

$ python --version
Python 2.6.6

$ ./TriBITS/tribits/python_utils/gitdist dist-repo-status
  File "./TriBITS/tribits/python_utils/gitdist", line 746
SyntaxError: Non-ASCII character '\xe2' in file ./TriBITS/tribits/python_utils/gitdist on line 746, but no encoding declared; see http://www.python.org/peps/pep-0263.html for details
```

Any idea what the problem might be?  Is this a Python version issue?

Also, can we add an option called something like `--dist-ascii-output` so that we can force ASCII output?  It is nice some times to produce ASCII-only output in order to copy and paste into a document or site that can only handle ASCII output.  And in that case, we will need unit tests for both ASCIII and the UTF-8 output where the ASCII output tests always get run on every platform.  Also, if there is a Python version issue, for Python 2.x it might just hard-code `--dist-ascii-output`.  Does that seam reasonable?

jmgate commented:
Yeah it very well might be a version issue.  I only develop on Python 3 so I never notice these things right off the bat.  I'll look into it, along with the `--dist-ascii-output` idea.

jmgate commented:
To give you an idea&hellip;

![screenshot from 2018-03-27 14-46-43](https://user-images.githubusercontent.com/20327215/37994099-b4b9bfd4-31cd-11e8-8688-96433fedf5a2.png)

jmgate commented:
Care to pull and try again, @bartlettroscoe?

jmgate commented:
Perhaps it's better to leave the default behavior as-is, and instead add a `--dist-utf8-output` flag to switch the output.

bartlettroscoe commented:
> Care to pull and try again, @bartlettroscoe?

I just tried again and got:

```
$ ./TriBITS/tribits/python_utils/gitdist dist-repo-status
Traceback (most recent call last):
  File "./TriBITS/tribits/python_utils/gitdist", line 1674, in <module>
    print(createTable(repoStatTable.getTableData()))
  File "./TriBITS/tribits/python_utils/gitdist", line 747, in createTable
    '┌─┬┐│┝━┿┥└┴┘'.encode(sys.stdout.encoding)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 0: ordinal not in range(128)
```

> Perhaps it's better to leave the default behavior as-is, and instead add a `--dist-utf8-output` flag to switch the output.

That might be a good idea.  And note that when you define your `gitdist-status` alias (or whatever you want to call it), you can define this as:

```
alias gitdist-status=gitdist --dist-utf8-output dist-repo-status
```

That way, you don't need to type anymore chars in your workflow.

Once concern will to ensure that this feature is protected.  We need to figure out how to get Travis CI to do a build and test of TriBITS and this script using a Python 3 env.

jmgate commented:
Yeah let me switch things around such that you need to request UTF-8 output, and I'll ping you when it's ready for another look.

jmgate commented:
Switched it over to require `--dist-utf8-output` if you want the pretty table.  Duplicated any unit tests that involved the table, so there are ASCII and UTF-8 versions of them.  Not sure how this'll work on Python 2 though.  Haven't done anything to default to ASCII if using Python 2.

bartlettroscoe commented:
@jmgate, thanks for making this change.  The only issue is that the unit tests crash with Python 2.6.6.   I submitted to CDash at:

* https://testing-vm.sandia.gov/cdash/testDetails.php?test=44715413&build=3397283

It shows the error:

```
================================================================================

TEST_0

Running: "/usr/bin/python" "/ascldap/users/rabartl/Trilinos.base/Trilinos/TriBITS/test/python_utils/gitdist_UnitTests.py" "-v"

  Running in working directory "/home/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/test/python_utils/TriBITS_gitdist_UnitTests"

--------------------------------------------------------------------------------

  File "/ascldap/users/rabartl/Trilinos.base/Trilinos/TriBITS/test/python_utils/gitdist_UnitTests.py", line 159
SyntaxError: Non-ASCII character '\xe2' in file /ascldap/users/rabartl/Trilinos.base/Trilinos/TriBITS/test/python_utils/gitdist_UnitTests.py on line 159, but no encoding declared; see http://www.python.org/peps/pep-0263.html for details

--------------------------------------------------------------------------------

TEST_0: Return code = 1
TEST_0: Pass criteria = Match REGEX {OK} [FAILED]
TEST_0: Result = FAILED

================================================================================
```

Can we protect the new unit tests to only run for Python 3.x?  Also, can we just hide the option `--dist-utf8-output` when the Python version is not 3.x?

bartlettroscoe commented:
@jmgate,

What is interesting is that this branch seems to pass with Python 2.7.6 which is used by the Travis CI build as shown at:

* https://my.cdash.org/index.php?project=TriBITS&parentid=1390940

with the unit test output shown at:

* https://my.cdash.org/testDetails.php?test=38579672&build=1390941

Can you see the new unit tests being run in that output?

jmgate commented:
I'm getting this when I try to view either of those CDash links:

<img width="487" alt="screen shot 2018-03-28 at 8 15 53 am" src="https://user-images.githubusercontent.com/20327215/38034835-56226cd0-3260-11e8-8a6f-614f3bdc9205.png">

Looks like roughly the same problem we're having with trilinos.org.

bartlettroscoe commented:
> I'm getting this when I try to view either of those CDash links

Yea, this is a known issue.  If you are not on the SRN network, you can get there just fine without a warning.  I am going to move over to use the CDash site at testing-vm.sandia.gov.  That is an updated CDash ite and it is open to everyone.

jmgate commented:
Mind trying one more time on 2.6.6?  Apparently specifying the file encoding doesn't change the fact that string literals are byte strings by default in Python 2.  This latest change should make it such that the appropriate strings are treated as Unicode strings.

bartlettroscoe commented:
I pulled (reset hard actually) the updated commit:

```
d4f406e "Use UTF-8 Box Drawing for dist-repo-status"
Author: Jason M. Gates <jmgate@sandia.gov>
Date:   Tue Mar 27 12:56:30 2018 -0600 (30 minutes ago)

M       test/python_utils/gitdist_UnitTests.py
M       tribits/ci_support/CheckinTest.py
M       tribits/ci_support/clone_extra_repos.py
M       tribits/python_utils/gitdist.py
```

and I am still getting:


```
210: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
210: 
210: Advanced Test: TriBITS_gitdist_UnitTests
210: 
210: Selected Test/CTest Propeties:
210:   CATEGORIES = BASIC
210:   PROCESSORS = 1
210:   TIMEOUT    = DEFAULT
210: 
210: Removing existing working directory '/home/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/test/python_utils/TriBITS_gitdist_UnitTests'
210: 
210: Creating new working directory '/home/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/test/python_utils/TriBITS_gitdist_UnitTests'
210: 
210: Running test commands: TEST_0
210: 
210: ================================================================================
210: 
210: TEST_0
210: 
210: Running: "/usr/bin/python" "/ascldap/users/rabartl/Trilinos.base/Trilinos/TriBITS/test/python_utils/gitdist_UnitTests.py" "-v"
210: 
210:   Running in working directory "/home/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/test/python_utils/TriBITS_gitdist_UnitTests"
210: 
210: --------------------------------------------------------------------------------
210: 
210:   File "/ascldap/users/rabartl/Trilinos.base/Trilinos/TriBITS/test/python_utils/gitdist_UnitTests.py", line 159
210: SyntaxError: Non-ASCII character '\xe2' in file /ascldap/users/rabartl/Trilinos.base/Trilinos/TriBITS/test/python_utils/gitdist_UnitTests.py on line 159, but no encoding declared; see http://www.python.org/peps/pep-0263.html for details
210: 
210: --------------------------------------------------------------------------------
210: 
210: TEST_0: Return code = 1
210: TEST_0: Pass criteria = Match REGEX {OK} [FAILED]
210: TEST_0: Result = FAILED
210: 
210: ================================================================================
210: 
210: OVERALL FINAL RESULT: TEST FAILED (TriBITS_gitdist_UnitTests)
210: 
210: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```

If it helps, you can log into my machine and debug this with Python 2.6.6  Contact me offline and I will provide simple instructions for reproducing this on my machine.

jmgate commented:
What a pain.  I'm just going to disable this for Python 2.

jmgate commented:
Right, disabled the UTF-8 output capability on Python 2.  The `--dist-utf8-output` option doesn't even show up.

jmgate commented on test/python_utils/gitdist_UnitTests.py:
```diff
@@ -125,17 +127,47 @@ def test_full_table(self):
+
+
+  def test_full_table_utf8(self):
+    if sys.version_info < (3,):
```
@bartlettroscoe, should this `if` be inside the function or outside, such that the function doesn't exist under Python 2?

bartlettroscoe replied:
@jmgate, I am not a Python expert but I think the if statement needs to be inside of the function.  I think what you have is fine because this function will never get called for Python versions 2.x.

bartlettroscoe commented:
@jmgate, have you given up on this PR?

jmgate commented:
Nope.  Just rebased on top of `master`.  We'll let TravisCI do it's thing, but then you can probably test it out and merge if it looks good to you.  I think I addressed all your concerns way back when&mdash;just haven't had a free moment to look at this in the last many months.

jmgate commented:
Note that this will conflict with #256.  Whichever one gets merged first, I can handle the merge conflicts and update the remaining one.

jmgate commented:
@bartlettroscoe, I took care of the merge conflicts that resulted from the merge of #256.  This should be good to go&mdash;both features are working together for me.

bartlettroscoe commented:
@jmgate, I will take a look after I get some other high-priority tasks done.

One thing we need to figure out is how to add python 3.x testing to the Travis CI testing to help support code like this.  Otherwise, I am fearful that it will break under future maintenance.

bartlettroscoe commented:
I pulled the branch:

```
$ cd Trilinos/TriBITS/

$ git fetch jmgate

$ git checkout --track jmgate/dist-repo-status-utf8-box-drawing

```

I then tested this first with Python 2.6.6 with:

```
$ which python
/usr/bin/python

$ python --version
Python 2.6.6

$ ./TriBITS/tribits/python_utils/gitdist dist-repo-status
-----------------------------------------------------------------------------------------------------------------------
| ID | Repo Dir            | Branch                            | Tracking Branch                          | C | M | ? |
|----|---------------------|-----------------------------------|------------------------------------------|---|---|---|
|  0 | Trilinos (Base)     | develop                           | github/develop                           |   |   |   |
|  1 | TriBITS             | dist-repo-status-utf8-box-drawing | jmgate/dist-repo-status-utf8-box-drawing |   |   |   |
|  2 | TriBITS/TriBITSDoc  | master                            | bb/master                                |   |   |   |
|  3 | kokkos              | develop                           | github/develop                           |   |   |   |
|  4 | kokkos/nvcc_wrapper | develop                           | github/develop                           |   |   |   |
|  5 | kokkos-kernels      | develop                           | github/develop                           |   |   |   |
|  6 | seacas              | master                            | gsjaardema/master                        |   |   |   |
-----------------------------------------------------------------------------------------------------------------------

(tip: to see a legend, pass in --dist-legend.)
```

I then tested this with Python sems-python/3.5.2:

```
$ module load sems-python/3.5.2 

$ which python3
/projects/sems/install/rhel6-x86_64/sems/compiler/python/3.5.2/bin/python3

$ python3 --version
Python 3.5.2

$ python3 ./TriBITS/tribits/python_utils/gitdist dist-repo-status  --dist-utf8-output
┌────┬─────────────────────┬───────────────────────────────────┬──────────────────────────────────────────┬───┬───┬───┐
│ ID │ Repo Dir            │ Branch                            │ Tracking Branch                          │ C │ M │ ? │
┝━━━━┿━━━━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┿━━━┿━━━┿━━━┥
│  0 │ Trilinos (Base)     │ develop                           │ github/develop                           │   │   │   │
│  1 │ TriBITS             │ dist-repo-status-utf8-box-drawing │ jmgate/dist-repo-status-utf8-box-drawing │   │   │   │
│  2 │ TriBITS/TriBITSDoc  │ master                            │ bb/master                                │   │   │   │
│  3 │ kokkos              │ develop                           │ github/develop                           │   │   │   │
│  4 │ kokkos/nvcc_wrapper │ develop                           │ github/develop                           │   │   │   │
│  5 │ kokkos-kernels      │ develop                           │ github/develop                           │   │   │   │
│  6 │ seacas              │ master                            │ gsjaardema/master                        │   │   │   │
└────┴─────────────────────┴───────────────────────────────────┴──────────────────────────────────────────┴───┴───┴───┘

(tip: to see a legend, pass in --dist-legend.)
```

@jmgate , for some reason, in my Linux terminal does not display the line between the column labels and the rows.  A screenshot is below.

![gitdist-utf-8-bix](https://user-images.githubusercontent.com/1959736/46081777-9d97f880-c15a-11e8-97aa-684965088f4e.png)

I am find merging this commit as it is, but it might be good to see why terminals like mine are not showing these uft-8 chars.

jmgate commented:
That's strange.  All the UTF-8 characters used are from the Box Drawing collection (0x2500&ndash;0x257F), so there's nothing special about that missing line of yours compared to the rest of them.

bartlettroscoe commented:
> What terminal are you using?

xterm running in Gnome 2.28.2 on a SNL COE RHEL6 machine.

bartlettroscoe commented:
Anyway, since this satisfies backward compatibility I will go a head and merge this branch.  Any issues with the `--dist-utf8-output` option can be fixed later.

Thanks for doing this!

jmgate commented:
Perhaps [this stackexchange thread](https://unix.stackexchange.com/questions/196152/xterm-not-displaying-unicode) can shed some light.  It looks like the issue is probably with xterm.
